pymc3/step_methods/atmcmc.py

```diff
@@ -0,0 +1,861 @@
+            progressbar=True, model=None, random_seed=None):
+
+    sampling = _iter_sample(draws, step, start, trace, chain,
+                            tune, model, random_seed)
+
+    if progressbar:
+        sampling = tqdm(sampling, total=draws)
+
+    try:
+        for strace in sampling:
+            pass
+
+    except KeyboardInterrupt:
+        pass
+
+    return chain
+
+
+def _iter_sample(draws, step, start=None, trace=None, chain=0, tune=None,
+                 model=None, random_seed=None):
+    """
+    Modified from :func:`pymc3.sampling._iter_sample` to be more efficient with
+    the ATMCMC algorithm.
+    """
+
+    model = modelcontext(model)
+
+    draws = int(draws)
+    seed(random_seed)
+    if draws < 1:
+        raise ValueError('Argument `draws` should be above 0.')
+
+    if start is None:
+        start = {}
+
+    point = pm.Point(start, model=model)
+    step.chain_index = chain
+
+    trace.setup(draws, chain)
+    for i in range(draws):
+        if i == tune:
+            step = pm.sampling.stop_tuning(step)
+
+        point, out_list = step.step(point)
+        trace.record(out_list)
+
+        yield trace
+
+
+def work_chain(work, pshared=None):
+    """
+    Wrapper function for parallel execution of _sample i.e. the Markov Chains.
+
+    Parameters
+    ----------
+    work : List
+        Containing all the information that is unique for each Markov Chain
+        i.e. [:class:'ATMCMC', chain_number(int),
+        sampling index(int), start_point(dictionary)]
+
+    Returns
+    -------
+    chain : int
+        Index of chain that has been sampled
+    """
+
+    if pshared is not None:
+        draws = pshared['draws']
+        progressbars = pshared['progressbars']
+        tune = pshared['tune']
+        trace_list = pshared['trace_list']
+        model = pshared['model']
+
+    step, chain, idx, start = work
+
+    progressbar = progressbars[idx]
+    trace = trace_list[idx]
+
+    return _sample(draws, step, start, trace, chain, tune, progressbar, model)
+
+
+def _iter_parallel_chains(draws, step, stage_path, progressbar, model, n_jobs,
+        chains=None):
+    """
+    Do Metropolis sampling over all the chains with each chain being
+    sampled 'draws' times. Parallel execution according to n_jobs.
+    """
+
+    if chains is None:
+        chains = list(range(step.n_chains))
+        idxs = chains
+    else:
+        idxs = list(range(len(chains)))
+
+    trace_list = []
+
+    if n_jobs > 1:
+        display = False
+
+    elif n_jobs == 1:
+        display = True
+
+    pack_pb = [progressbar for i in range(n_jobs - 1)] + [display]
+    block_pb = []
+    list_pb = []
+
+    for i in range(int(len(chains) / n_jobs)):
+        block_pb.append(pack_pb)
+
+    map(list_pb.extend, block_pb)
+
+    pm._log.info('Initialising chain traces ...')
+    for chain in chains:
+        trace_list.append(atext.Text(stage_path, model=model))
+
+    pm._log.info('Sampling ...')
+
+    pshared = dict(
```

Comment:
you can put whatever is meant not to be forked, so this differs for each function that is supposed to be parallelised, so if I understood correct this wont work with a named tuple ...